# 

Dr. Rao is studying an endangered wolf whose population is declining by 15% every year. If there are currently 5,100 wolves, how many will remain in 6 years?

The function describing the number of wolves after n years can be given by
<br>
5100 * (85 / 100)^n
<br>
= 5100 * (17 / 20)^n

In [1]:
from sympy import Symbol, Function, Eq, solve
import matplotlib.pyplot as plt

In [2]:
n = Symbol('n') # Number of years since the current time
w = Function('w')(n) # Number of wolves after n years
eq = Eq(w, 5100 * (17 / 20)**n) # Equation for number of wolves after n years
eq

Eq(w(n), 5100*0.85**n)

In [3]:
eq.subs({n : 6}) # Getting the number of wolves that will be left after 6 years

Eq(w(6), 1923.4625296875)

After 6 years, we can estimate there to be around 1923 wolves, given the current rate of decline. Note that we don't use the exponential decay model here, because that model is used when the rate of change declines as the target variable's value declines. In this casem it would be used if the wolf population's decline would slow down in proportion to the decrease in its population.

# 

In medieval times there were 10,000 people living in a city that was struck by a plague so that people began to die at an exponential rate daily. After 6 days, there were only 8,500 people living. How many were living after three weeks?

## Explanation

I first thought that people dying at an exponential rate meant that poeple died in an exponentially increasing rate, meaning that the rate of decline in population rose with time. Hence, to model the number of the dead, we could use exponential growth model.
<br><br>
However, the question may mean that the people died at an exponentially decreasing rate, which is what would happen in the case of population decline, since fewer the people, the less the chances of any one person contracting the disease from other people, and hence, the smaller the number of fatalities.
<br><br>
I will work with both interpretations, first with people dying at an exponentially increasing rate, then people dying at an exponentially decreasing rate.
<br><br>

## Growth model on the number of dead people

The exponential growth model (to model the number of total dead people) is given by
<br>
$\frac{d m(t)}{dt} = k m(t)$
<br>
where m(t) is the number of dead people after t days, and k is a constant that determines the rate at which the rate of change increases.

In [17]:
from sympy import Symbol, Function, Eq, solve, dsolve, diff
import matplotlib.pyplot as plt
import numpy as np

In [77]:
t = Symbol('t') # Number of days since the plague first hit
m = Function('m')(t) # Number of dead people after t days
k = Symbol('k') # Proportional constant
C1 = Symbol('C1') # Potential integration constant
de = Eq(m.diff(t), k * m) # The differential equation mentioned above
de

Eq(Derivative(m(t), t), k*m(t))

In [39]:
ge = dsolve(de) # General equation derived from the above differential equation
ge

Eq(m(t), C1*exp(k*t))

The first day i.e. t = 0 is the day when the first person dies. Hence,
<br>
m(0) = 1
<br>
Given this, we substitute these values in the general equation and solve for C1. We can see that
<br>
C1 = 1
<br>
but for the sake of establishing a generalised process, we shall solve it regardless.

In [40]:
solve(ge.subs({m : 1, t : 0}))

[1]

In [41]:
pe = ge.subs({C1 : 1})
pe

Eq(m(t), exp(k*t))

To find k, we substitute
<br>
t = 6 and m = 10000 - 8500
<br>
since in the data we have that after 6 days since the first death, there are 8500 people left, meaning
<br>
10000 - 8500
<br>
people have died since. For this substitution, we solve for k.

In [42]:
solve(pe.subs({m : 10000 - 8500, t : 6}), k)

[log(2**(1/3)*3**(1/6)*sqrt(5)) - 2*I*pi/3,
 log(2**(1/3)*3**(1/6)*sqrt(5)) - I*pi/3,
 log(2**(1/3)*3**(1/6)*sqrt(5)) + I*pi/3,
 log(2**(1/3)*3**(1/6)*sqrt(5)) + 2*I*pi/3,
 log(2**(1/3)*3**(1/6)*sqrt(5)) + I*pi,
 log(2**(1/3)*3**(1/6)*sqrt(5))]

We chose the only real valued solution here, since complex solutions are not applicable in concrete populations.

In [45]:
kvalue = np.log(2**(1/3) * 3**(1/6) * np.sqrt(5))
pe = pe.subs({k : kvalue})
pe

Eq(m(t), exp(1.21887006451505*t))

Three weeks amounts to 7 * 3 = 21 days. Applying this to our newly found equation, we get the following...

In [46]:
pe.subs({t : 21})

Eq(m(21), 130713187934.501)

This would mean that there is no one left by the end of three weeks.

## Decay model for number of living people

Now, working with the second interpretation, that the population is decreasing in an exponentially decreasing rate.
<br><br>
Here, the population of living people is decreasing, so that will be the target variable. The exponential decay model (to model the number of total living people left) is given by
<br>
$\frac{d a(t)}{dt} = -k a(t)$
<br>
where a(t) is the number of alive people after t days, and -k is a constant that determines the rate at which the rate of change decreases.

In [68]:
a = Function('a')(t) # Number of dead people after t days
de = Eq(a.diff(t), -k * a) # The differential equation mentioned above
de

Eq(Derivative(a(t), t), -k*a(t))

In [69]:
ge = dsolve(de) # General equation derived from the above differential equation
ge

Eq(a(t), C1*exp(-k*t))

The first day i.e. t = 0 is the day when the the population is intact. Hence,
<br>
m(0) = 10000
<br>
Given this, we substitute these values in the general equation and solve for C1. We can see that
<br>
C1 = 10000
<br>
but for the sake of establishing a generalised process, we shall solve it regardless.

In [72]:
solve(ge.subs({a : 10000, t : 0}), C1)

[10000]

In [73]:
pe = ge.subs({C1 : 10000})
pe

Eq(a(t), 10000*exp(-k*t))

To find k, we substitute
<br>
t = 6 and m = 8500
<br>
since in the data we have that after 6 days since the first death, there are 8500 people left. For this substitution, we solve for k.

In [75]:
solve(pe.subs({a : 8500, t : 6}), k)

[log(17**(5/6)*2**(1/3)*5**(1/6)/17) - 2*I*pi/3,
 log(17**(5/6)*2**(1/3)*5**(1/6)/17) - I*pi/3,
 log(17**(5/6)*2**(1/3)*5**(1/6)/17) + I*pi/3,
 log(17**(5/6)*2**(1/3)*5**(1/6)/17) + 2*I*pi/3,
 log(17**(5/6)*2**(1/3)*5**(1/6)/17) + I*pi,
 log(17**(5/6)*2**(1/3)*5**(1/6)/17)]

We chose the only real valued solution here, since complex solutions are not applicable in concrete populations.

In [65]:
kvalue = np.log(17**(5/6) * 2**(1/3) * 5**(1/6)/17)
pe = pe.subs({k : kvalue})
pe

Eq(a(t), 10000*exp(-0.0270864882496291*t))

Three weeks amounts to 7 * 3 = 21 days. Applying this to our newly found equation, we get the following...

In [66]:
pe.subs({t : 21})

Eq(a(21), 5661.952739835)

Based on this model, after three weeks, around 5662 people are left in the city.

## Notes

The meanings of the first day i.e. t = 0 differed based on what would allow the model to work at all. In the growth model for dead people, if we started with t = 0 meaning no death occurs, we would get C1 = 0, which will not lead to any further meaningful results. Similarly, for the decay model for living people, if we started with t = 0 meaning the first death occurs, we will not be able to solve for C1. Note that different interpretations of t = 0 does not change the predictive power of the model.

# 

A scientist started with a culture of 10 bacteria in a dish. The number of bacteria at the end of each successive minute increased exponentially so that the number at the end of one hour was 500. How many hours (to the nearest hour) from the start of the experiment passed before there were 1,000,000 bacteria?

This is a clear case of exponential growth, which is modelled by the following equation...
<br>
$\frac{d b(t)}{dt} = k b(t)$
<br>
where b(t) is the number of bacteria after t minutes, and k is a constant that determines the rate at which the rate of change increases.

In [109]:
from sympy import Symbol, Function, Eq, solve, dsolve, diff
import matplotlib.pyplot as plt
import numpy as np

In [95]:
t = Symbol('t') # Number of minutes since the 10 bacteria were initially observed
b = Function('b')(t) # Number of bacteria after t minutes
k = Symbol('k') # Proportional constant
C1 = Symbol('C1') # Potential integration constant
de = Eq(b.diff(t), k * b) # The differential equation mentioned above
de

Eq(Derivative(b(t), t), k*b(t))

In [96]:
ge = dsolve(de) # General equation derived from the above differential equation
ge

Eq(b(t), C1*exp(k*t))

The first minute i.e. t = 0 is the minute when 10 bacteria are initially observed. Hence,
<br>
b(0) = 10
<br>
Given this, we substitute these values in the general equation and solve for C1. We can see that
<br>
C1 = 10
<br>
but for the sake of establishing a generalised process, we shall solve it regardless.

In [97]:
solve(ge.subs({b : 10, t : 0}))

[10]

In [102]:
pe = ge.subs({C1 : 10})
pe

Eq(b(t), 10*exp(k*t))

To find k, we substitute
<br>
t = 60 (one hour) and b = 500
<br>
since in the data we have that after one hour, there are 500 bacteria. For this substitution, we solve for k.

In [103]:
solve(pe.subs({b : 500, t : 60}), k)

[log(2**(1/60)*5**(1/30)) - 9*I*pi/10,
 log(2**(1/60)*5**(1/30)) - 5*I*pi/6,
 log(2**(1/60)*5**(1/30)) - 7*I*pi/10,
 log(2**(1/60)*5**(1/30)) - 2*I*pi/3,
 log(2**(1/60)*5**(1/30)) - I*pi/2,
 log(2**(1/60)*5**(1/30)) - I*pi/3,
 log(2**(1/60)*5**(1/30)) - 3*I*pi/10,
 log(2**(1/60)*5**(1/30)) - I*pi/6,
 log(2**(1/60)*5**(1/30)) - I*pi/10,
 log(2**(1/60)*5**(1/30)) + I*pi/10,
 log(2**(1/60)*5**(1/30)) + I*pi/6,
 log(2**(1/60)*5**(1/30)) + 3*I*pi/10,
 log(2**(1/60)*5**(1/30)) + I*pi/3,
 log(2**(1/60)*5**(1/30)) + I*pi/2,
 log(2**(1/60)*5**(1/30)) + 2*I*pi/3,
 log(2**(1/60)*5**(1/30)) + 7*I*pi/10,
 log(2**(1/60)*5**(1/30)) + 5*I*pi/6,
 log(2**(1/60)*5**(1/30)) + 9*I*pi/10,
 log(2**(1/60)*5**(1/30)) + I*pi,
 log(2**(1/60)*5**(1/30)),
 log(-2**(1/60)*5**(1/30)/4 + 2**(1/60)*5**(8/15)/4 - 2**(1/60)*5**(1/30)*I*sqrt(sqrt(5)/8 + 5/8)),
 log(-2**(1/60)*5**(1/30)/4 + 2**(1/60)*5**(8/15)/4 + 2**(1/60)*5**(1/30)*I*sqrt(sqrt(5)/8 + 5/8)),
 log(2**(1/60)*5**(1/30)/4 + 2**(1/60)*5**(8/15)/4 - 2**(1/60)*

We chose the only real valued solution here, since complex solutions are not applicable in concrete populations.

In [104]:
kvalue = np.log(2**(1/60)*5**(1/30))
pe_actual = pe.subs({k : kvalue})
pe_actual

Eq(b(t), 10*exp(k*t))

Now, putting b as 1000000 and solving for t may not be something the computer can solve quickly. Hence, we shall solve for
<br>
$b(t) = 10e^x$
<br>
where $x = 0.0652003834238025t$
<br>
The simplest way to achieve this is to solve for t with k = 1. This will not be the actual t value, but rather, will be what x was supposed to be... It may be a little confusing, but it is the simplest way to get the result practically.

In [107]:
kvalue = 1
pe_modified = pe.subs({k : kvalue})
pe_modified

Eq(b(t), 10*exp(t))

Substituting b = 1000000, we solve for t in the modified equation...

In [112]:
solve(pe_modified.subs({b : 1000000}), t)

[log(100000)]

In this modified equation, we know that t is actually x, where $x = 0.0652003834238025t$. Hence, we have that
<br>
$log(1000000) = 0.0652003834238025t$
<br>
Hence, the actual value of t is $t = \frac{log(1000000)}{0.0652003834238025}$

In [114]:
np.log(1000000) / 0.0652003834238025

211.8930876244002

In hours...

In [115]:
211.8930876244002 / 60

3.53155146040667

After around 212 minutes, or around 3.6 hours, there will be more than 500 bacteria.